network/crl_test.go

```diff
@@ -0,0 +1,468 @@
+
+// Convert PEM-encoded certificate into DER and parse it into x509.Certificate
+func pemToX509Cert(data []byte) (*x509.Certificate, error) {
+	certDERBlock, _ := pem.Decode(data)
+	if certDERBlock == nil {
+		return nil, errors.New("Failed to decode PEM")
+	}
+
+	if certDERBlock.Type != "CERTIFICATE" {
+		return nil, errors.New("Decoded DER block was not a certificate")
+	}
+
+	cert, err := x509.ParseCertificate(certDERBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+// Starts HTTP server on some random port, uses `mux` to handle requests,
+// returns created server and IP:port of listening socket
+func getTestHTTPServer(t *testing.T, mux *http.ServeMux) (*http.Server, string) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("Cannot create TCP socket for HTTP server: %v", err)
+	}
+
+	addr := listener.Addr().String()
+
+	httpServer := &http.Server{
+		Handler: mux,
+	}
+	go func() {
+		httpServer.Serve(listener)
+	}()
+
+	// Give the server time to start, otherwise this test will fail sometimes
+	time.Sleep(time.Millisecond * 100)
+
+	return httpServer, addr
+}
+
+func getTestChain(t *testing.T, prefix string, paths ...string) ([][]byte, [][]*x509.Certificate) {
+	if len(paths) < 2 {
+		t.Fatal("Certificate chain should contain at least two certificates\n")
+	}
+
+	rawCerts := [][]byte{}
+
+	verifiedChains := [][]*x509.Certificate{}
+	verifiedChains = append(verifiedChains, []*x509.Certificate{})
+
+	// Read and parse certificates, construct the chain
+	for _, p := range paths {
+		filename := path.Join(prefix, p)
+
+		rawCert, err := ioutil.ReadFile(filename)
+		if err != nil {
+			t.Fatalf("Cannot read certificate: %v\n", err)
+		}
+
+		cert, err := pemToX509Cert(rawCert)
+		if err != nil {
+			t.Fatalf("Cannot parse certificate: %v", err)
+		}
+
+		rawCerts = append(rawCerts, rawCert)
+		verifiedChains[0] = append(verifiedChains[0], cert)
+	}
+
+	// Make sure this function was called with valid arguments
+	for i, cert := range verifiedChains[0] {
+		if i == 0 && cert.IsCA {
+			// Avoid situations where we got two CA certificates,
+			// first one should be used for authentication/encryption
+			t.Fatalf("First certificate in chain (`%s`) should not be CA", cert.Subject.String())
+		}
+
+		var nextCert *x509.Certificate
+		if i < len(verifiedChains[0])-1 {
+			nextCert = verifiedChains[0][i+1]
+		}
+
+		if nextCert != nil {
+			// Every certificate (except last) should be signed by the next one
+			err := cert.CheckSignatureFrom(nextCert)
+			if err != nil {
+				t.Fatalf("Certificate `%s` was not signed by `%s`: %v\n", cert.Subject.String(), nextCert.Subject.String(), err)
+			}
+		}
+	}
+
+	return rawCerts, verifiedChains
+}
+
+func getValidTestChain(t *testing.T) ([][]byte, [][]*x509.Certificate) {
+	return getTestChain(
+		t,
+		TestCertPrefix,
+		TestCertFilename,
+		TestCACertFilename)
+}
+
+// func getValidTestChain3(t *testing.T) ([][]byte, [][]*x509.Certificate) {
```

Comment:
same about commented functions

Reply:
As the certificate chain with intermediate CA is added again, these lines are now uncommented